criu/pidfd.c

```diff
@@ -0,0 +1,257 @@
+	for (int i = 0; i < DEAD_PIDFD_HASH_SIZE; i++)
+		INIT_HLIST_HEAD(&dead_pidfd_hash[i]);
+}
+
+static struct dead_pidfd *lookup_dead_pidfd(unsigned int ino)
+{
+	struct dead_pidfd *dead;
+	struct hlist_head *chain;
+
+	chain = &dead_pidfd_hash[ino % DEAD_PIDFD_HASH_SIZE];
+	hlist_for_each_entry(dead, chain, hash) {
+		if (dead->ino == ino)
+			return dead;
+	}
+	return NULL;
+}
+
+static void free_dead_pidfd_hash(void)
+{
+	int i;
+	struct dead_pidfd *dead;
+	struct hlist_node *tmp;
+
+	for (i = 0; i < DEAD_PIDFD_HASH_SIZE; i++) {
+		hlist_for_each_entry_safe(dead, tmp, &dead_pidfd_hash[i], hash) {
+			xfree(dead);
+		}
+		INIT_HLIST_HEAD(&dead_pidfd_hash[i]);
+	}
+}
+
+int is_pidfd_link(char *link)
+{
+	/*
+	* pidfs was introduced in Linux 6.9
+	* before which anonymous-inodes were used
+	*/
+	return is_anon_link_type(link, "[pidfd]");
+}
+
+static void pr_info_pidfd(char *action, PidfdEntry *pidfe)
+{
+	pr_info("%s: id %#08x flags %u NSpid %d ino %u\n",
+		action, pidfe->id, pidfe->flags, pidfe->nspid, pidfe->ino
+	);
+}
+
+static int dump_one_pidfd(int pidfd, u32 id, const struct fd_parms *p)
+{
+	struct pidfd_dump_info pidfd_info = {.pidfe = PIDFD_ENTRY__INIT};
+	FileEntry fe = FILE_ENTRY__INIT;
+
+	if (parse_fdinfo(pidfd, FD_TYPES__PIDFD, &pidfd_info))
+		return -1;
+
+	if (p->flags & PIDFD_THREAD) {
+		pr_err("PIDFD_THREAD flag is currently not supported\n");
+		return -1;
+	}
+
+	/*
+	* Check if the pid pidfd refers to is part of process tree
+	* This ensures the process will exist on restore.
+	*/
+	if (pidfd_info.pid != -1 && !pstree_item_by_real(pidfd_info.pid)) {
+		pr_err("pidfd pid %d is not a part of process tree..\n",
+			pidfd_info.pid);
+		return -1;
+	}
+
+	pidfd_info.pidfe.id = id;
+	pidfd_info.pidfe.flags = (p->flags & ~O_RDWR);
+	pidfd_info.pidfe.fown = (FownEntry *)&p->fown;
+
+	fe.type = FD_TYPES__PIDFD;
+	fe.id = pidfd_info.pidfe.id;
+	fe.pidfd = &pidfd_info.pidfe;
+
+	pr_info_pidfd("Dumping", &pidfd_info.pidfe);
+	return pb_write_one(img_from_set(glob_imgset, CR_FD_FILES), &fe, PB_FILE);
+}
+
+const struct fdtype_ops pidfd_dump_ops = {
+	.type = FD_TYPES__PIDFD,
+	.dump = dump_one_pidfd,
+};
+
+static int pidfd_open(pid_t pid, int flags)
+{
+	return syscall(__NR_pidfd_open, pid, flags);
+}
+
+static int create_tmp_process(void)
+{
+	int tmp_process;
+	tmp_process = fork();
+	if (tmp_process < 0) {
+		pr_perror("Could not fork");
+		return -1;
+	} else if (tmp_process == 0) {
+		while(1)
+			sleep(1);
+	}
+	return tmp_process;
+}
+
+static int open_one_pidfd(struct file_desc *d, int *new_fd)
+{
+	struct pidfd_info *info;
+	struct dead_pidfd *dead;
+	int pidfd, status;
+
+	info = container_of(d, struct pidfd_info, d);
+	if (info->pidfe->nspid != -1) {
+		pidfd = pidfd_open(info->pidfe->nspid, info->pidfe->flags);
+		if (pidfd < 0) {
+			pr_perror("Could not open pidfd for %d", info->pidfe->nspid);
+			goto err_close;
+		}
+		goto out;
+	}
+
+	dead = lookup_dead_pidfd(info->pidfe->ino);
+	BUG_ON(!dead);
+
+	dead->count--;
```

Comment:
`dead` isn't shared between processes. If I don't miss something it can be updated from multiple processes. You need allocate these entries from a shared pool and to use some form of synchronization to ensure data consistency and prevent race conditions.

Reply:
ah, nice catch! Thanks!

@bsach64 you need to use `shmalloc` function (just grep over the CRIU codebase). Also, we have a flag `COLLECT_SHARED` you can pay attention to a different places in the CRIU code how we use it.

Idea is that you need to pre-allocate all the needed data-structures in a shared memory (usually, on the image collect stage before we start fork-ing processes), so then you can use this memory to pass data between different processes.

Feel free to reach me if you have any questions.

Reply:
Okay, Alex!
I will get back to you in a couple of days!

Reply:
Hello @avagin!
With a lot of help from alex, I think we have resolved this problem!